plugins/base/src/main/kotlin/generation/SingleModuleGeneration.kt

```diff
@@ -61,11 +63,18 @@ public class SingleModuleGeneration(private val context: DokkaContext) : Generat
 
     override val generationName: String = "documentation for ${context.configuration.moduleName}"
 
-    public fun createDocumentationModels(): List<DModule> = runBlocking(Dispatchers.Default) {
-        context.configuration.sourceSets.parallelMap { sourceSet -> translateSources(sourceSet, context) }.flatten()
-            .also { modules -> if (modules.isEmpty()) exitGenerationGracefully("Nothing to document") }
+    /**
+     * Implementation note: it runs in a separated single thread due to existing support of coroutines (see #2936)
+     */
+    @OptIn(DelicateCoroutinesApi::class)
+    public fun createDocumentationModels(): List<DModule> = newSingleThreadContext("Generating documentable model").use { coroutineContext -> // see https://github.com/Kotlin/dokka/issues/3151
+        runBlocking(coroutineContext) {
+            context.configuration.sourceSets.parallelMap { sourceSet -> translateSources(sourceSet, context) }.flatten()
+                .also { modules -> if (modules.isEmpty()) exitGenerationGracefully("Nothing to document") }
+        }
     }
```

Comment:
There's a considerable chance it can lead to noticeably worse performance in projects with K1, and it might be especially evident with the CLI runner. Maybe you had time to do a few test runs to compare before and after?

It breaks the abstraction, but what do you think about using this single-threaded implementation for K2 only? Just to avoid raising too many questions for now with nothing to offer

Reply:
Update on this question after some internal discussions:

Vadim did some benchmarking of the CLI runner to see if this has any impact on performance, and it looks like the documentable model takes 20% more time to be generated, but it's just a fraction of the overall time Dokka takes to generate documentation, so it's not really visible if you look at the whole run and not just an individual step.

Since we plan to get rid of coroutines either way, it's an acceptable solution for now.